kie-camel/src/main/java/org/kie/camel/KieProducer.java

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.kie.camel;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.stream.Collector;
+import java.util.stream.Stream;
+
+import org.apache.camel.Exchange;
+import org.apache.camel.impl.DefaultProducer;
+import org.kie.api.KieServices;
+import org.kie.api.command.Command;
+import org.kie.server.api.model.ServiceResponse;
+import org.kie.server.client.KieServicesClient;
+import org.kie.server.client.KieServicesFactory;
+import org.kie.server.client.ProcessServicesClient;
+import org.kie.server.client.RuleServicesClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static java.util.stream.Collectors.groupingBy;
+import static org.kie.camel.KieCamelUtils.getResultMessage;
+import static org.kie.camel.KieComponent.KIE_CLIENT;
+import static org.kie.camel.KieComponent.KIE_OPERATION;
+
+public class KieProducer extends DefaultProducer {
+
+    private static final transient Logger log = LoggerFactory.getLogger( KieProducer.class );
+
+    private static final String DEFAULT_CLIENT = "KieServices";
+
+    private static final String KIE_HEADERS_PREFIX = "kie.";
+
+    private final KieServicesClient client;
+
+    private final Map<String, InternalProducer> producers = new HashMap<>();
+
+    public KieProducer( KieEndpoint endpoint ) {
+        super(endpoint);
+        client = KieServicesFactory.newKieServicesClient( endpoint.getConfiguration() );
+    }
+
+    @Override
+    public void process( Exchange exchange ) throws Exception {
+        writeResponse( exchange, getProducer(exchange).execute(exchange) );
+    }
+
+    private void writeResponse( Exchange exchange, Object response ) {
```

Comment:
wouldn't it be better to move that method (or it's responsibility) into producer? then we would not have to deal with service response if/else and the process method above would simply call:

```
getProducer(exchange).execute(exchange)
```